### ISSUE TITLE
reconnect on network down

### ISSUE THREAD

**Issue by z0mb1ek** [redisson/redisson#946]
Hi. Debugging my swarm cluster and down my network, then I see:
`Message: org.redisson.client.RedisTimeoutException: Redis server response timeout (500 ms) occured for command: (INCRBY) with params: [..., 0] channel: [id: 0x315e345a, L:/172.18.0.8:54874 - R:/192.168.167.182:6379].`
Only restart help me. Why can you reconnect?

**Comment by jackygurui** [comment]
This exception only happens after retry

**Comment by z0mb1ek** [comment]
no,there was many exceptions,and no up, only reboot can help me

**Comment by jackygurui** [comment]
Redisson should connect to your next master once the role transition is over

**Comment by z0mb1ek** [comment]
its a single node mode, there was no reconnect

**Comment by jackygurui** [comment]
Can you show us your configuration and usage please? Also what's the version you're having this problem?

**Comment by z0mb1ek** [comment]
I have latest version, my config is: 

```
Config config = new Config();
        config
            .setCodec(new StringCodec())
            .useSingleServer()
            .setAddress("redis://" + host + ":" + port)
            .setRetryAttempts(1)
            .setRetryInterval(0)
            .setTimeout(timeout);
        return Redisson.create(config);
```
I have docker environment, when my redis down and start on other node, I see timeouts on redisson client, seems like it recconected not all pool to new redis. Restart solve problem. very sad(

**Comment by johnou** [comment]
Why are you using useSingleServer with multiple nodes? What's the TTL of the host? Are you running a security manager in your JVM?

**Comment by z0mb1ek** [comment]
there is 1 node!!!there is no security manager

**Comment by z0mb1ek** [comment]
I see this after 6 hours, I think that not all pool change dns for new redis instance.

**Comment by mrniko** [comment]
Here is my answer for similar case https://github.com/redisson/redisson/issues/914#issuecomment-314416749

**Comment by z0mb1ek** [comment]
@mrniko 127.0.0.1??? it is not test for my use case, I told you that dns resolve has bug.

**Comment by mrniko** [comment]
@z0mb1ek 
> I told you that dns resolve has bug.

Which bug, could you help to resolve it?

**Comment by mrniko** [comment]
there is an option `Config.dnsMonitoring` have you tried it?

**Comment by z0mb1ek** [comment]
@mrniko it turn on by default I see. I try to debug it and report. But I see this on 10 apps in my cluster. The same behavior at all

**Comment by mrniko** [comment]
@z0mb1ek 

Thank you! I'm looking forward for your results

**Comment by z0mb1ek** [comment]
@mrniko 
This code does not reconnect =(

```java
Config config = new Config();
        config
            .setCodec(new StringCodec())
            .useSingleServer()
            .setAddress("redis://" + "localhost" + ":" + 6379)
            .setRetryAttempts(1)
            .setRetryInterval(0);
        RedissonClient redissonClient = Redisson.create(config);

        for (int i = 0; i < 1_000_000_000; i++) {
            try {
                Thread.sleep(500);
                redissonClient.getBucket("test").set("1");
            } catch (Exception e) {
                System.out.println(e.toString());
            }
        }
```

**Comment by wsz87** [comment]
It seems that I have exactly the same problem. The following code and logs may help you with finding the cause of the problem:

Configuration of a Redisson client:
```
final Config redissonConfig = new Config();
        redissonConfig.setNettyThreads(2);
        redissonConfig.useSingleServer()
                .setAddress("redis://elasticache-single-node.cache.amazonaws.com:6379")
                .setConnectionPoolSize(10)
                .setConnectionMinimumIdleSize(1)
                .setIdleConnectionTimeout(5000)
                .setReconnectionTimeout(3000)
                .setTimeout(5000)
                .setPingTimeout(2000)
                .setClientName("DefaultSyncStateCacheClient");
```

The client is created only once and shared across many sessions:
```
        final RedissonClient redissonClient = Redisson.create(redissonConfig);
        redissonClient.getNodesGroup().addConnectionListener(new RedisConnectionListener());
```

The listener is configured to notify about connection status:
```
public class RedisConnectionListener implements ConnectionListener {

    private static final Logger LOGGER = LoggerFactory.getLogger(RedisConnectionListener.class);

    @Override
    public void onConnect(final InetSocketAddress address) {
        LOGGER.info("Connecting to: [hostString={}]", address.getHostString());
    }

    @Override
    public void onDisconnect(final InetSocketAddress address) {
        LOGGER.info("Disconnecting from: [hostString={}]", address.getHostString());
    }
}
```

There is also a health check implemented to check status of connection and it is invoked with GET /admin/healthcheck:
```
public class LockHealthCheck extends HealthCheck {

    static final String NAME = "lockState";

    private final RedissonClient redissonClient;

    @Inject
    public LockHealthCheck(final RedissonClient redissonClient) {
        super();

        this.redissonClient = redissonClient;
    }

    @Override
    protected Result check() {
        if (redissonClient.getNodesGroup().pingAll()) {
            // returnsOK (200)
            return Result.healthy("All nodes pinged successfully");
        }

        // return INTERNAL_SERVER_ERROR (500)
        return Result.unhealthy("At least one node did not answer to ping correctly in specified time");
    }
}
```

Below is the code which accesses RLocalCachedMap (the map instance in Redis has the same name). "isMapEntryPresent" method is invoked with POST /application/getMapEntry.
```
class TestReconnection {

    private final RLocalCachedMap<String, String> map = redissonClient.getLocalCachedMap(
                "RLocalCachedMap",
                LocalCachedMapOptions.<String, String>defaults()
                        .invalidationPolicy(LocalCachedMapOptions.InvalidationPolicy.ON_CHANGE_WITH_CLEAR_ON_RECONNECT)
                        .maxIdle(5, TimeUnit.SECONDS)

    public boolean isMapEntryPresent() {
        return Optional.ofNullable(map.get("MAP_ENTRY_KEY"))
                       .map(Objects::nonNull)
                       .orElse(false);
    }
}
```

NOTE: the code is not 100 % equal to the real code, but the conception and configuration is exact.

I've got a Redis server in Elasticache service which has following parameters:
- tcp-keepalive: 300 (I don't think it is relevant here)
- timeout: 30

This completes the setup description. Now I reproduced network downtime and reconnection with following steps:

Firstly, I pinged all nodes with my health check (3 times).
```
0:0:0:0:0:0:0:1 [2017-08-01 11:41:54.916] null "GET /admin/healthcheck HTTP/1.1" 200 130 96 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/59.0.3071.115 Safari/537.36"
0:0:0:0:0:0:0:1 [2017-08-01 11:42:00.850] null "GET /admin/healthcheck HTTP/1.1" 200 130 96 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/59.0.3071.115 Safari/537.36"
0:0:0:0:0:0:0:1 [2017-08-01 11:42:02.556] null "GET /admin/healthcheck HTTP/1.1" 200 130 95 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/59.0.3071.115 Safari/537.36"
```

And as we can see pingAll returned true for all three times.

Secondly, I turned on my WiFi network. And I started seeing these logs repeatedly:
```
ERROR  [2017-08-01 13:42:18.521]  [/] [redisson-3-5] SingleConnectionManager - elasticache-single-node.cache.amazonaws.com: nodename nor servname provided, or not known 
java.net.UnknownHostException: elasticache-single-node.cache.amazonaws.com: nodename nor servname provided, or not known
	at java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method)
	at java.net.InetAddress$2.lookupAllHostAddr(InetAddress.java:928)
	at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1323)
	at java.net.InetAddress.getAllByName0(InetAddress.java:1276)
	at java.net.InetAddress.getAllByName(InetAddress.java:1192)
	at java.net.InetAddress.getAllByName(InetAddress.java:1126)
	at java.net.InetAddress.getByName(InetAddress.java:1076)
	at org.redisson.connection.SingleConnectionManager$1$1.run(SingleConnectionManager.java:105)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
ERROR  [2017-08-01 13:42:23.526]  [/] [redisson-3-6] SingleConnectionManager - elasticache-single-node.cache.amazonaws.com 
java.net.UnknownHostException: elasticache-single-node.cache.amazonaws.com
	at java.net.InetAddress.getAllByName0(InetAddress.java:1280)
	at java.net.InetAddress.getAllByName(InetAddress.java:1192)
	at java.net.InetAddress.getAllByName(InetAddress.java:1126)
	at java.net.InetAddress.getByName(InetAddress.java:1076)
	at org.redisson.connection.SingleConnectionManager$1$1.run(SingleConnectionManager.java:105)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
ERROR  [2017-08-01 13:42:28.528]  [/] [redisson-3-7] SingleConnectionManager - elasticache-single-node.cache.amazonaws.com: nodename nor servname provided, or not known 
java.net.UnknownHostException: elasticache-single-node.cache.amazonaws.com: nodename nor servname provided, or not known
	at java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method)
	at java.net.InetAddress$2.lookupAllHostAddr(InetAddress.java:928)
	at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1323)
	at java.net.InetAddress.getAllByName0(InetAddress.java:1276)
	at java.net.InetAddress.getAllByName(InetAddress.java:1192)
	at java.net.InetAddress.getAllByName(InetAddress.java:1126)
	at java.net.InetAddress.getByName(InetAddress.java:1076)
	at org.redisson.connection.SingleConnectionManager$1$1.run(SingleConnectionManager.java:105)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
ERROR  [2017-08-01 13:42:33.530]  [/] [redisson-3-8] SingleConnectionManager - elasticache-single-node.cache.amazonaws.com 
java.net.UnknownHostException: elasticache-single-node.cache.amazonaws.com
	at java.net.InetAddress.getAllByName0(InetAddress.java:1280)
	at java.net.InetAddress.getAllByName(InetAddress.java:1192)
	at java.net.InetAddress.getAllByName(InetAddress.java:1126)
	at java.net.InetAddress.getByName(InetAddress.java:1076)
	at org.redisson.connection.SingleConnectionManager$1$1.run(SingleConnectionManager.java:105)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
ERROR  [2017-08-01 13:42:38.533]  [/] [redisson-3-1] SingleConnectionManager - elasticache-single-node.cache.amazonaws.com: nodename nor servname provided, or not known 
java.net.UnknownHostException: elasticache-single-node.cache.amazonaws.com: nodename nor servname provided, or not known
	at java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method)
	at java.net.InetAddress$2.lookupAllHostAddr(InetAddress.java:928)
	at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1323)
	at java.net.InetAddress.getAllByName0(InetAddress.java:1276)
	at java.net.InetAddress.getAllByName(InetAddress.java:1192)
	at java.net.InetAddress.getAllByName(InetAddress.java:1126)
	at java.net.InetAddress.getByName(InetAddress.java:1076)
	at org.redisson.connection.SingleConnectionManager$1$1.run(SingleConnectionManager.java:105)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
ERROR  [2017-08-01 13:42:43.533]  [/] [redisson-3-2] SingleConnectionManager - elasticache-single-node.cache.amazonaws.com 
java.net.UnknownHostException: elasticache-single-node.cache.amazonaws.com
	at java.net.InetAddress.getAllByName0(InetAddress.java:1280)
	at java.net.InetAddress.getAllByName(InetAddress.java:1192)
	at java.net.InetAddress.getAllByName(InetAddress.java:1126)
	at java.net.InetAddress.getByName(InetAddress.java:1076)
	at org.redisson.connection.SingleConnectionManager$1$1.run(SingleConnectionManager.java:105)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
ERROR  [2017-08-01 13:42:48.537]  [/] [redisson-3-3] SingleConnectionManager - elasticache-single-node.cache.amazonaws.com: nodename nor servname provided, or not known 
java.net.UnknownHostException: elasticache-single-node.cache.amazonaws.com: nodename nor servname provided, or not known
	at java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method)
	at java.net.InetAddress$2.lookupAllHostAddr(InetAddress.java:928)
	at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1323)
	at java.net.InetAddress.getAllByName0(InetAddress.java:1276)
	at java.net.InetAddress.getAllByName(InetAddress.java:1192)
	at java.net.InetAddress.getAllByName(InetAddress.java:1126)
	at java.net.InetAddress.getByName(InetAddress.java:1076)
	at org.redisson.connection.SingleConnectionManager$1$1.run(SingleConnectionManager.java:105)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
ERROR  [2017-08-01 13:42:53.542]  [/] [redisson-3-4] SingleConnectionManager - elasticache-single-node.cache.amazonaws.com 
java.net.UnknownHostException: elasticache-single-node.cache.amazonaws.com
	at java.net.InetAddress.getAllByName0(InetAddress.java:1280)
	at java.net.InetAddress.getAllByName(InetAddress.java:1192)
	at java.net.InetAddress.getAllByName(InetAddress.java:1126)
	at java.net.InetAddress.getByName(InetAddress.java:1076)
	at org.redisson.connection.SingleConnectionManager$1$1.run(SingleConnectionManager.java:105)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
ERROR  [2017-08-01 13:42:58.548]  [/] [redisson-3-5] SingleConnectionManager - elasticache-single-node.cache.amazonaws.com: nodename nor servname provided, or not known 
java.net.UnknownHostException: elasticache-single-node.cache.amazonaws.com: nodename nor servname provided, or not known
	at java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method)
	at java.net.InetAddress$2.lookupAllHostAddr(InetAddress.java:928)
	at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1323)
	at java.net.InetAddress.getAllByName0(InetAddress.java:1276)
	at java.net.InetAddress.getAllByName(InetAddress.java:1192)
	at java.net.InetAddress.getAllByName(InetAddress.java:1126)
	at java.net.InetAddress.getByName(InetAddress.java:1076)
	at org.redisson.connection.SingleConnectionManager$1$1.run(SingleConnectionManager.java:105)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
ERROR  [2017-08-01 13:43:03.551]  [/] [redisson-3-6] SingleConnectionManager - elasticache-single-node.cache.amazonaws.com 
java.net.UnknownHostException: elasticache-single-node.cache.amazonaws.com
	at java.net.InetAddress.getAllByName0(InetAddress.java:1280)
	at java.net.InetAddress.getAllByName(InetAddress.java:1192)
	at java.net.InetAddress.getAllByName(InetAddress.java:1126)
	at java.net.InetAddress.getByName(InetAddress.java:1076)
	at org.redisson.connection.SingleConnectionManager$1$1.run(SingleConnectionManager.java:105)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
ERROR  [2017-08-01 13:43:08.558]  [/] [redisson-3-7] SingleConnectionManager - elasticache-single-node.cache.amazonaws.com: nodename nor servname provided, or not known 
java.net.UnknownHostException: elasticache-single-node.cache.amazonaws.com: nodename nor servname provided, or not known
	at java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method)
	at java.net.InetAddress$2.lookupAllHostAddr(InetAddress.java:928)
	at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1323)
	at java.net.InetAddress.getAllByName0(InetAddress.java:1276)
	at java.net.InetAddress.getAllByName(InetAddress.java:1192)
	at java.net.InetAddress.getAllByName(InetAddress.java:1126)
	at java.net.InetAddress.getByName(InetAddress.java:1076)
	at org.redisson.connection.SingleConnectionManager$1$1.run(SingleConnectionManager.java:105)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
ERROR  [2017-08-01 13:43:13.559]  [/] [redisson-3-8] SingleConnectionManager - elasticache-single-node.cache.amazonaws.com 
java.net.UnknownHostException: elasticache-single-node.cache.amazonaws.com
	at java.net.InetAddress.getAllByName0(InetAddress.java:1280)
	at java.net.InetAddress.getAllByName(InetAddress.java:1192)
	at java.net.InetAddress.getAllByName(InetAddress.java:1126)
	at java.net.InetAddress.getByName(InetAddress.java:1076)
	at org.redisson.connection.SingleConnectionManager$1$1.run(SingleConnectionManager.java:105)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
ERROR  [2017-08-01 13:43:18.565]  [/] [redisson-3-1] SingleConnectionManager - elasticache-single-node.cache.amazonaws.com: nodename nor servname provided, or not known 
java.net.UnknownHostException: elasticache-single-node.cache.amazonaws.com: nodename nor servname provided, or not known
	at java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method)
	at java.net.InetAddress$2.lookupAllHostAddr(InetAddress.java:928)
	at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1323)
	at java.net.InetAddress.getAllByName0(InetAddress.java:1276)
	at java.net.InetAddress.getAllByName(InetAddress.java:1192)
	at java.net.InetAddress.getAllByName(InetAddress.java:1126)
	at java.net.InetAddress.getByName(InetAddress.java:1076)
	at org.redisson.connection.SingleConnectionManager$1$1.run(SingleConnectionManager.java:105)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
ERROR  [2017-08-01 13:43:23.569]  [/] [redisson-3-2] SingleConnectionManager - elasticache-single-node.cache.amazonaws.com 
java.net.UnknownHostException: elasticache-single-node.cache.amazonaws.com
	at java.net.InetAddress.getAllByName0(InetAddress.java:1280)
	at java.net.InetAddress.getAllByName(InetAddress.java:1192)
	at java.net.InetAddress.getAllByName(InetAddress.java:1126)
	at java.net.InetAddress.getByName(InetAddress.java:1076)
	at org.redisson.connection.SingleConnectionManager$1$1.run(SingleConnectionManager.java:105)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
ERROR  [2017-08-01 13:43:28.576]  [/] [redisson-3-3] SingleConnectionManager - elasticache-single-node.cache.amazonaws.com: nodename nor servname provided, or not known 
java.net.UnknownHostException: elasticache-single-node.cache.amazonaws.com: nodename nor servname provided, or not known
	at java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method)
	at java.net.InetAddress$2.lookupAllHostAddr(InetAddress.java:928)
	at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1323)
	at java.net.InetAddress.getAllByName0(InetAddress.java:1276)
	at java.net.InetAddress.getAllByName(InetAddress.java:1192)
	at java.net.InetAddress.getAllByName(InetAddress.java:1126)
	at java.net.InetAddress.getByName(InetAddress.java:1076)
	at org.redisson.connection.SingleConnectionManager$1$1.run(SingleConnectionManager.java:105)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
ERROR  [2017-08-01 13:43:33.577]  [/] [redisson-3-4] SingleConnectionManager - elasticache-single-node.cache.amazonaws.com 
java.net.UnknownHostException: elasticache-single-node.cache.amazonaws.com
	at java.net.InetAddress.getAllByName0(InetAddress.java:1280)
	at java.net.InetAddress.getAllByName(InetAddress.java:1192)
	at java.net.InetAddress.getAllByName(InetAddress.java:1126)
	at java.net.InetAddress.getByName(InetAddress.java:1076)
	at org.redisson.connection.SingleConnectionManager$1$1.run(SingleConnectionManager.java:105)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
ERROR  [2017-08-01 13:43:38.584]  [/] [redisson-3-5] SingleConnectionManager - elasticache-single-node.cache.amazonaws.com: nodename nor servname provided, or not known 
java.net.UnknownHostException: elasticache-single-node.cache.amazonaws.com: nodename nor servname provided, or not known
	at java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method)
	at java.net.InetAddress$2.lookupAllHostAddr(InetAddress.java:928)
	at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1323)
	at java.net.InetAddress.getAllByName0(InetAddress.java:1276)
	at java.net.InetAddress.getAllByName(InetAddress.java:1192)
	at java.net.InetAddress.getAllByName(InetAddress.java:1126)
	at java.net.InetAddress.getByName(InetAddress.java:1076)
	at org.redisson.connection.SingleConnectionManager$1$1.run(SingleConnectionManager.java:105)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
ERROR  [2017-08-01 13:43:43.585]  [/] [redisson-3-6] SingleConnectionManager - elasticache-single-node.cache.amazonaws.com 
java.net.UnknownHostException: elasticache-single-node.cache.amazonaws.com
	at java.net.InetAddress.getAllByName0(InetAddress.java:1280)
	at java.net.InetAddress.getAllByName(InetAddress.java:1192)
	at java.net.InetAddress.getAllByName(InetAddress.java:1126)
	at java.net.InetAddress.getByName(InetAddress.java:1076)
	at org.redisson.connection.SingleConnectionManager$1$1.run(SingleConnectionManager.java:105)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
``` 

Thirdly, I turned on my WiFi network (after over 60 seconds to make sure that Redis server close idle connections - timeout is configured to 30sec) and after a short while it stopped writing above logs.
Fourthly, I started to check whether I can access map entry with 'isMapEntryPresent' method:
```
0:0:0:0:0:0:0:1 [2017-08-01 11:44:14.587] nZLhaXQUWU-bs-mb "OPTIONS /application/getMapEntry HTTP/1.1" 200 24 195 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/59.0.3071.115 Safari/537.36"
ERROR  [2017-08-01 13:44:21.169] 1GYByXsrO4-bs-mb [MAP_ENTRY_KEY/wojciech] [dw-28 - POST /application/getMapEntry] RuntimeExceptionMapper - Runtime exception caught. 
org.redisson.client.RedisTimeoutException: Redis server response timeout (5000 ms) occured for command: (HGET) with params: [RLocalCachedMap, MAP_ENTRY_KEY] channel: [id: 0xf027d973, L:/192.168.15.199:53351 - R:elasticache-single-node.cache.amazonaws.com/10.0.0.1:6379]
	at org.redisson.command.CommandAsyncService$10.run(CommandAsyncService.java:646)
	at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:663)
	at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:738)
	at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:466)
	at java.lang.Thread.run(Thread.java:748)
0:0:0:0:0:0:0:1 [2017-08-01 11:44:21.203] 1GYByXsrO4-bs-mb "POST /application/getMapEntry HTTP/1.1" 500 285 6610 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/59.0.3071.115 Safari/537.36"

ERROR  [2017-08-01 13:44:36.256] o9IxDBzOWH-bs-mb [MAP_ENTRY_KEY/wojciech] [dw-22 - POST /application/getMapEntry] RuntimeExceptionMapper - Runtime exception caught. 
org.redisson.client.RedisTimeoutException: Command execution timeout for command: (HGET) with params: [RLocalCachedMap, MAP_ENTRY_KEY]
	at org.redisson.command.CommandAsyncService$7.run(CommandAsyncService.java:537)
	at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:663)
	at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:738)
	at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:466)
	at java.lang.Thread.run(Thread.java:748)
0:0:0:0:0:0:0:1 [2017-08-01 11:44:36.260] o9IxDBzOWH-bs-mb "POST /application/getMapEntry HTTP/1.1" 500 132 6324 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/59.0.3071.115 Safari/537.36"


ERROR  [2017-08-01 13:44:52.555] GlXZsCQSbm-bs-mb [MAP_ENTRY_KEY/wojciech] [dw-24 - POST /application/getMapEntry] RuntimeExceptionMapper - Runtime exception caught. 
org.redisson.client.RedisTimeoutException: Command execution timeout for command: (HGET) with params: [RLocalCachedMap, MAP_ENTRY_KEY]
	at org.redisson.command.CommandAsyncService$7.run(CommandAsyncService.java:537)
	at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:663)
	at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:738)
	at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:466)
	at java.lang.Thread.run(Thread.java:748)
0:0:0:0:0:0:0:1 [2017-08-01 11:44:52.558] GlXZsCQSbm-bs-mb "POST /application/getMapEntry HTTP/1.1" 500 132 6412 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/59.0.3071.115 Safari/537.36"


ERROR  [2017-08-01 13:45:08.052] EyjPsKioBi-bs-mb [MAP_ENTRY_KEY/wojciech] [dw-28 - POST /application/getMapEntry] RuntimeExceptionMapper - Runtime exception caught. 
org.redisson.client.RedisTimeoutException: Command execution timeout for command: (HGET) with params: [RLocalCachedMap, MAP_ENTRY_KEY]
	at org.redisson.command.CommandAsyncService$7.run(CommandAsyncService.java:537)
	at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:663)
	at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:738)
	at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:466)
	at java.lang.Thread.run(Thread.java:748)
0:0:0:0:0:0:0:1 [2017-08-01 11:45:08.055] EyjPsKioBi-bs-mb "POST /application/getMapEntry HTTP/1.1" 500 132 6354 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/59.0.3071.115 Safari/537.36"



ERROR  [2017-08-01 13:45:26.053] KTC3iDGgoq-bs-mb [MAP_ENTRY_KEY/wojciech] [dw-22 - POST /application/getMapEntry] RuntimeExceptionMapper - Runtime exception caught. 
org.redisson.client.RedisTimeoutException: Command execution timeout for command: (HGET) with params: [RLocalCachedMap, MAP_ENTRY_KEY]
	at org.redisson.command.CommandAsyncService$7.run(CommandAsyncService.java:537)
	at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:663)
	at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:738)
	at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:466)
	at java.lang.Thread.run(Thread.java:748)
0:0:0:0:0:0:0:1 [2017-08-01 11:45:26.055] KTC3iDGgoq-bs-mb "POST /application/getMapEntry HTTP/1.1" 500 132 6409 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/59.0.3071.115 Safari/537.36"


ERROR  [2017-08-01 13:45:40.655] 5iwxm35geY-bs-mb [MAP_ENTRY_KEY/wojciech] [dw-24 - POST /application/getMapEntry] RuntimeExceptionMapper - Runtime exception caught. 
org.redisson.client.RedisTimeoutException: Command execution timeout for command: (HGET) with params: [RLocalCachedMap, MAP_ENTRY_KEY]
	at org.redisson.command.CommandAsyncService$7.run(CommandAsyncService.java:537)
	at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:663)
	at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:738)
	at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:466)
	at java.lang.Thread.run(Thread.java:748)
  
0:0:0:0:0:0:0:1 [2017-08-01 11:45:40.657] 5iwxm35geY-bs-mb "POST /application/getMapEntry HTTP/1.1" 500 132 6358 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/59.0.3071.115 Safari/537.36"
``` 

After all those checks there was an error:
```
ERROR  [2017-08-01 13:45:48.930]  [/] [redisson-netty-1-2] CommandsQueue - Exception occured. Channel: [id: 0xf027d973, L:/192.168.15.199:53351 - R:elasticache-single-node.cache.amazonaws.com/10.0.0.1:6379] 
java.io.IOException: Operation timed out
	at sun.nio.ch.FileDispatcherImpl.read0(Native Method)
	at sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39)
	at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:223)
	at sun.nio.ch.IOUtil.read(IOUtil.java:192)
	at sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:380)
	at io.netty.buffer.PooledUnsafeDirectByteBuf.setBytes(PooledUnsafeDirectByteBuf.java:288)
	at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1100)
	at io.netty.channel.socket.nio.NioSocketChannel.doReadBytes(NioSocketChannel.java:372)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:123)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:644)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:579)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:496)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:458)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
```

And it started working again:
```
INFO   [2017-08-01 13:45:48.931]  [/] [redisson-netty-1-2] RedisConnectionListener - Disconnecting from: [hostString=elasticache-single-node.cache.amazonaws.com] 
INFO   [2017-08-01 13:45:49.012]  [/] [redisson-netty-1-2] RedisConnectionListener - Connecting to: [hostString=elasticache-single-node.cache.amazonaws.com] 
INFO   [2017-08-01 13:46:19.277]  [/] [redisson-netty-1-2] RedisConnectionListener - Disconnecting from: [hostString=elasticache-single-node.cache.amazonaws.com] 
INFO   [2017-08-01 13:46:19.412]  [/] [redisson-netty-1-2] RedisConnectionListener - Connecting to: [hostString=elasticache-single-node.cache.amazonaws.com] 

0:0:0:0:0:0:0:1 [2017-08-01 11:46:38.501] 9Q9p7U6nUJ-bs-mb "POST /application/getMapEntry HTTP/1.1" 201 75 309 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/59.0.3071.115 Safari/537.36"
0:0:0:0:0:0:0:1 [2017-08-01 11:46:42.323] HK0PblXzrB-bs-mb "POST /application/getMapEntry HTTP/1.1" 201 75 57 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/59.0.3071.115 Safari/537.36"

INFO   [2017-08-01 13:47:13.282]  [/] [redisson-netty-1-2] RedisConnectionListener - Disconnecting from: [hostString=elasticache-single-node.cache.amazonaws.com] 
INFO   [2017-08-01 13:47:13.414]  [/] [redisson-netty-1-2] RedisConnectionListener - Connecting to: [hostString=elasticache-single-node.cache.amazonaws.com] 

0:0:0:0:0:0:0:1 [2017-08-01 11:47:31.499] Jf0s0MoTpZ-bs-mb "POST /application/getMapEntry HTTP/1.1" 201 75 48 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/59.0.3071.115 Safari/537.36"
0:0:0:0:0:0:0:1 [2017-08-01 11:47:39.982] zXnZEXwyXk-bs-mb "POST /application/getMapEntry HTTP/1.1" 201 75 47 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/59.0.3071.115 Safari/537.36"

INFO   [2017-08-01 13:48:10.207]  [/] [redisson-netty-1-2] RedisConnectionListener - Disconnecting from: [hostString=elasticache-single-node.cache.amazonaws.com] 
INFO   [2017-08-01 13:48:10.327]  [/] [redisson-netty-1-2] RedisConnectionListener - Connecting to: [hostString=elasticache-single-node.cache.amazonaws.com] 
INFO   [2017-08-01 13:48:41.721]  [/] [redisson-netty-1-2] RedisConnectionListener - Disconnecting from: [hostString=elasticache-single-node.cache.amazonaws.com] 
INFO   [2017-08-01 13:48:41.811]  [/] [redisson-netty-1-2] RedisConnectionListener - Connecting to: [hostString=elasticache-single-node.cache.amazonaws.com] 
```
The last two logs were repeating continuously.
The weird thing for me I didn't have those logs before and they stopped occurring with exception:
```
WARN   [2017-08-01 14:50:15.107]  [/] [redisson-netty-1-1] ConnectionWatchdog - Can't connect RedisConnection@481029080 [redisClient=[addr=elasticache-single-node.cache.amazonaws.com/10.0.0.1:6379], channel=[id: 0x7f3a3320, L:0.0.0.0/0.0.0.0:53777]] to elasticache-single-node.cache.amazonaws.com/10.0.0.1:6379 
org.redisson.client.RedisTimeoutException: Command execution timeout for elasticache-single-node.cache.amazonaws.com/10.0.0.1:6379
	at org.redisson.client.RedisConnection$2.run(RedisConnection.java:212)
	at io.netty.util.concurrent.PromiseTask$RunnableAdapter.call(PromiseTask.java:38)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:120)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:462)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
```
The exception happened after my computer fell asleep.

**Comment by wsz87** [comment]
I think it would be best to implement some sort of keep alive (tcp level or Redis PING called repeatedly) feature as Redisson / Netty is probably unaware of problem with connections as they are closed from the other side and Redisson / Netty were not present at time of disconnection. I've tried to find an option that may control the behaviour of connections, but I haven't found it. I think it may affect in-use and idle (including those from minimal idle pool) connections.

**Comment by z0mb1ek** [comment]
any news? catch it again(

**Comment by mrniko** [comment]
@wsz87 @z0mb1ek I have created a separate issue for this PING feature request https://github.com/redisson/redisson/issues/1088

**Comment by tobad357** [comment]
So we can still replicate this issue quite easily
When running redis behind nat and killing the server that redis is on which causes  netty to still think the link is up.

redisson --- NAT IP --- redis

then kill redis server

redisson -- NAT IP xxx redis

This causes netty to still think redis is up and in turn redisson will throw Timeout exceptions
even if redis is up again as a reconnect doesnt happen

An example of this is if running redis in kubernetes

**Comment by tobad357** [comment]
Seems like the new Ping feature isn't working for us in the NAT case.

**Comment by johnou** [comment]
@mrniko is keep alive configured in Netty? ChannelOption.SO_KEEPALIVE, true

**Comment by mrniko** [comment]
@johnou

Seems Redisson connected to some middle host not to Redis host directly.

**Comment by johnou** [comment]
@mrniko I think regardless this should be considered for applications connecting to Redis behind NAT..

```java
    private Bootstrap createBootstrap(RedisClientConfig config, Type type) {
        Bootstrap bootstrap = new Bootstrap()
                        .channel(config.getSocketChannelClass())
                        .group(config.getGroup())
                        .remoteAddress(addr);

        bootstrap.handler(new RedisChannelInitializer(bootstrap, config, this, channels, type));
        bootstrap.option(ChannelOption.CONNECT_TIMEOUT_MILLIS, config.getConnectTimeout());
        bootstrap.option(ChannelOption.SO_KEEPALIVE, true); // **************
        return bootstrap;
    }
```

**Comment by johnou** [comment]
aha cool, it has already been added in a later version of Redisson.

```java
        bootstrap.option(ChannelOption.SO_KEEPALIVE, config.isKeepAlive());
```

@tobad357 try to enable keepalive and see how that goes..

```java
ReplicatedServersConfig clusterConfig = config.useReplicatedServers(); // your server config might be instantiated differently.
clusterConfig.setKeepAlive(true);
```

**Comment by tobad357** [comment]
We tried this
```
useSingleServer with.setPingConnection(true)
            .setKeepAlive(true)
            .setTcpNoDelay(true)
```

And still get the same issue

**Comment by mrniko** [comment]
Seems Redisson connected to some middle host not to Redis host directly.
@tobad357 Am I right?

**Comment by tobad357** [comment]
@mrniko basically how Kubernetes works is that redis is behind NAT
When you deploy a container/pod to kubernetes it gets a IP on the POD network and then a stable IP on a service network

When a pod gets redeployed it gets a new IP but the SVC IP stays the same.
Kubernetes then uses IPTABLES to NAT the connection between SVC IP and the POD IP
The IPTABLES will be updated if POD IP changes

So technically it is as if Redisson is connecting to a FW who in turn forwards the traffic to redis.

**Comment by mrniko** [comment]
I have introduced `pingConnectionInterval` setting. Please use it
